### DISCUSSION TITLE
blog/array-types-in-type-script

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#98] [Announcements]
# blog/array-types-in-type-script

is string[] really better than Array<string> ? Of course not!

https://tkdodo.eu/blog/array-types-in-type-script

**Comment by hudochenkov** [comment]
Great argumentation!

P. S. It was hard to read the article because spaces were missing after dots in so many sentences. And for some reason, there is almost always a space before question marks at the end of sentences. Also spaces are missing before many inline code fragments.

**Reply by TkDodo** [reply]
Sorry for that, I guess I was in a hurry publishing it today. If you want to do me a favour, you can open a PR to fix the ones you found: https://github.com/tkdodo/blog

Thanks

**Comment by jaworek** [comment]
Thanks for this article! I preferred using ts-eslint rule "@typescript-eslint/array-type" with option set to "array-simple" because as you mentioned it is shorter. It avoid most of the issues that are mentioned here, but I guess it wouldn't hurt much to have it longer, but consistent throughout the codebase.

**Reply by TkDodo** [reply]
Yeah that setting is an okay middle ground, as you said, it avoids those issues.

**Comment by Murkrage** [comment]
This makes total sense. I ran into some of these situations before so definitely considering doing this in my own codebases.

**Comment by Artaud** [comment]
Thanks for the article! Frankly, the only argument that convinces me is the readonly issue where readonly and array def is separated. 
The others are not as convincing to me. Especially the parentheses just feel natural to include in this case and I never had any issue with them.
Regarding readability, you argue that we say "this is an array of strings", while I am used to saying "this is a string array" (maybe due to coming from Java world), so string[] comes more naturally.

**Reply by Artaud** [reply]
And while I agree that having to look at a possible [] at the end of an object can be a source of errors, the compiler still has your back in that case and it comes down to being used to check the [].

**Reply by TkDodo** [reply]
My background is also in the world of JVM languages, but if you look into the direction where Kotlin and Scala are heading with collections, it's `List<String>`.

> Especially the parentheses just feel natural to include in this case and I never had any issue with them.

Then I think you're more the exception than the rule. When I'm working on a team where we might also have Juniors onboard, I'd like to avoid as much chance for confusing / errors as possible.

**Reply by ImSejin** [reply]
@TkDodo

Absolutely agree.

**Comment by Tiwari007** [comment]
I Just started with typescript. and sometimes face the same thing so thanks for this well written docs. 
It really helped 🙌🙌

**Comment by vulkanosaure** [comment]
Maybe a 2nd argument for the type[] notations (to balance it a bit more) 
It works better for 2D array (or more)

Array<Array<string>> 

But while I agree that keeping the code short is not the number 1 priority, in that case, I still consider having a short syntax (for something that you'd write everywhere) a more important win than all the arguments you mentionned

**Reply by vulkanosaure** [reply]
The 2D array like got broken : 
Array < Array < type> > 
Is a pain both to write and to read

**Comment by carlossuds** [comment]
Would have to add as const to make it work: `pick(myObject, ['foo', 'bar'] as const)`

**Comment by Maximilian94** [comment]
Thanks for the article !!

**Comment by hoanguye99** [comment]
Great article!

**Comment by lishine** [comment]
More in favor of the short form:
I will leave the Helper<xxx> to typescript helpers.
Have you have the generic notation, you are looking for angle brackets to see where it starts and ends - I don't like it.
Having both notations in a code make the developer more get used to reading whatever code there is. 
And yes for me short is important. This is part of the reason I prefer let...

**Comment by RuinousCheng** [comment]
If possible，I'd like to write it as `[]string` or `[]<string>`.

**Comment by uy-scuti100** [comment]
wow, this really opened my eyes to so many things about typescript. This is Gold

**Comment by wm6692** [comment]
The year was one thousand nine hundred and eighty five Anno Domini. Marty time traveled to one thousand nine hundred and fifty five Anno Domini.

   or

1985 -> 1955 

It's not just about being short, it's about the conventions we use.
miles vs km
why 12 hours on a clock and not 10?
why 60 minutes in an hour and not 10 or 100? 
why 360 degrees in a circle and not 100 or 1000?

**Comment by eranhirsch** [comment]
My main problem with Array<T> is that it doesn't play nicely with tuples, which can only use the former notation (e.g. `readonly [string, boolean]`). Because tuples are JS arrays under the hood, using Array<T> would mean that we have 2 different ways to define the same kind of JS object.

e.g. a regular readonly array would be notated as `ReadonlyArray<T>`, but a readonly non-empty array would be `readonly [T, ...Array<T>]`... the difference in my mind is too big. I'd rather the notation be more similar (e.g. `readonly T[]`, and `readonly [T, ...T[]]`)

**Reply by TkDodo** [reply]
The difference to me is that with tuples, the notation starts with `[]` while for the array syntax, it comes at the end. That's why I don't mind it for tuples (it's the only possibility as you said). I also don't mind the difference that much, but I can see why this could be an argument for `T[]`

**Comment by maksnester** [comment]
On one of the past projects we had a eslint rule that for simple types enforced `foo[]` syntax and for more complex ones `Array<ComplexStuff<Bar>>`
I think it makes sense as `string[]` is easy to read and `ComplexStuff<Bar>[]` is absolutely not.

**Reply by maksnester** [reply]
The rule https://typescript-eslint.io/rules/array-type/


I would just use array-simple as the only option for the rule

[
  {
    /** The array type expected for mutable cases. */
    default?: 'array-simple';
  },
];

Disagree with the author that `string[]` or even `readonly string[]` is hard to read, it's not 🙃
